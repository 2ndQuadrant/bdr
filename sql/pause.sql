\ccc regression

SELECT bdr.bdr_apply_is_paused();

SELECT bdr.bdr_replicate_ddl_command('CREATE TABLE public.pause_test(x text primary key);');
INSERT INTO pause_test(x) VALUES ('before pause');
SELECT pg_xlog_wait_remote_apply(pg_current_xlog_location(), 0);

\cc postgres

SELECT bdr.bdr_apply_is_paused();
SELECT bdr.bdr_apply_pause();
SELECT bdr.bdr_apply_is_paused();
-- It's necessary to wait for a latch timeout on apply workers
-- until bdr_apply_pause gets taught to set their latches.
SELECT pg_sleep(6);

\ccc regression

INSERT INTO pause_test(x) VALUES ('after pause before resume');

\cc postgres

-- Give more time for a row to replicate if it's going to
-- (it shouldn't)
SELECT pg_sleep(1);

-- Pause state is preserved across sessions
SELECT bdr.bdr_apply_is_paused();

-- Must not see row from after pause
SELECT x FROM pause_test;

SELECT bdr.bdr_apply_resume();

\ccc regression

INSERT INTO pause_test(x) VALUES ('after resume');

-- The pause latch timeout is 5 minutes. To make sure that setting
-- the latch is doing its job and unpausing before timeout, expect
-- resume to take effect well before then.
BEGIN;
SET LOCAL statement_timeout = '60s';
SELECT pg_xlog_wait_remote_apply(pg_current_xlog_location(), 0);
COMMIT;

\cc postgres

-- Must see all three rows
SELECT x FROM pause_test;
