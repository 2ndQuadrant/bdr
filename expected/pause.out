\ccc regression
invalid command \ccc
SELECT bdr.bdr_apply_is_paused();
 bdr_apply_is_paused 
---------------------
 f
(1 row)

SELECT bdr.bdr_replicate_ddl_command('CREATE TABLE public.pause_test(x text primary key);');
 bdr_replicate_ddl_command 
---------------------------
 
(1 row)

INSERT INTO pause_test(x) VALUES ('before pause');
SELECT pg_xlog_wait_remote_apply(pg_current_xlog_location(), 0);
 pg_xlog_wait_remote_apply 
---------------------------
 
(1 row)

\cc postgres
invalid command \cc
SELECT bdr.bdr_apply_is_paused();
 bdr_apply_is_paused 
---------------------
 f
(1 row)

SELECT bdr.bdr_apply_pause();
 bdr_apply_pause 
-----------------
 
(1 row)

SELECT bdr.bdr_apply_is_paused();
 bdr_apply_is_paused 
---------------------
 t
(1 row)

-- It's necessary to wait for a latch timeout on apply workers
-- until bdr_apply_pause gets taught to set their latches.
SELECT pg_sleep(6);
 pg_sleep 
----------
 
(1 row)

\ccc regression
invalid command \ccc
INSERT INTO pause_test(x) VALUES ('after pause before resume');
\cc postgres
invalid command \cc
-- Give more time for a row to replicate if it's going to
-- (it shouldn't)
SELECT pg_sleep(1);
 pg_sleep 
----------
 
(1 row)

-- Pause state is preserved across sessions
SELECT bdr.bdr_apply_is_paused();
 bdr_apply_is_paused 
---------------------
 t
(1 row)

-- Must not see row from after pause
SELECT x FROM pause_test;
             x             
---------------------------
 before pause
 after pause before resume
(2 rows)

SELECT bdr.bdr_apply_resume();
 bdr_apply_resume 
------------------
 
(1 row)

\ccc regression
invalid command \ccc
INSERT INTO pause_test(x) VALUES ('after resume');
-- The pause latch timeout is 5 minutes. To make sure that setting
-- the latch is doing its job and unpausing before timeout, expect
-- resume to take effect well before then.
BEGIN;
SET LOCAL statement_timeout = '60s';
SELECT pg_xlog_wait_remote_apply(pg_current_xlog_location(), 0);
 pg_xlog_wait_remote_apply 
---------------------------
 
(1 row)

COMMIT;
\cc postgres
invalid command \cc
-- Must see all three rows
SELECT x FROM pause_test;
             x             
---------------------------
 before pause
 after pause before resume
 after resume
(3 rows)

